### ISSUE TITLE
ConnectionInterface and their implementations

### ISSUE THREAD

**Issue by NoelDeMartin** [laravel/ideas#1221]
After some frustrating experiences working with existing mongodb packages for Laravel, I decided to dip my toes on a test project and implement mongodb integration from scratch. And just from the get go, I found some design incongruencies that I think should be fixed, regardless of mongodb or nosql support in general. Seeing this I also suspect the same problem may be present in other parts of the code, so I want to discuss this before looking elsewhere.

The problem I found is the following. There is currently an interface under the Database namespace called `ConnectionInterface`. To anyone who wants to implement a new connection, I assume this is the interface to implement. However, after implementing this interface I got an error coming from the `DatabaseManager`, because it has different methods that take a `Connection` instance, not a `ConnectionInterface`. So I went ahead and also re-implemented the database manager, using the interface `ConnectionResolverInterface`. This was not ideal, but still correct from a design perspective. But then I found the first incorrect scenario. When running migrations, the class `DatabaseMigrationRepository` is used. This class takes as an argument in the constructor a `ConnectionResolverInterface`, however `$this->getConnection()->getSchemaBuilder();` is called in multiple ocasions. This is not correct, because the object returned by `getConnection` is a `ConnectionInterface`, which lacks the `getSchemaBuilder` method. In order to fix this, either the `ConnectionResolverInterface` should be modified to return a `Connection` or the `ConnectionInterface` should be modified to include the `getSchemaBuilder` method.

I guess you can already see the problem here. I won't go further than this particular example, but this same problem is present in other places. This effectively makes the `ConnectionInterface` useless because you need to end up extending the `Connection` class itself in most cases.

I am aware that this does not have an easy fix, because modifying interfaces has the potential to break a lot of existing code. But I think this should be fixed at some point, or it has the potential to become worse with time.

**Comment by RemiCollin** [comment]
I agree these should be addressed. If it's potential breaking change, we can still PR targeting 5.7. Did you check how existing laravel database libraries workaround this issue ?

**Comment by NoelDeMartin** [comment]
Yeah, the couple I have seen extend `Connection` and base classes directly, ignoring the interfaces altogether.

**Comment by RemiCollin** [comment]
I wonder if this could be by design. `DatabaseManager` seems to be tightly bound to `Connection` class, and `ConnectionInterface` is barely used outside of the `QueryBuilder`. 

This means to me, that, theoretically, you could use the `QueryBuilder` outside of Laravel with any `ConnectionInterface` implementation. However, if you want to add a DB driver to laravel, there is no interface for it and you have to extends `Illuminate\Database\Connection` 

Altogether, I guess the name `ConnectionInterface` is misleading; should have been name something along `DatabaseQueryInterface`.

**Comment by NoelDeMartin** [comment]
Yes but that's what I am trying to say, by design that is not correct. If the intention is what you mentioned, then the dependency on the constructor of `DatabaseMigrationRepository` should be a `DatabaseManager`, not a `ConnectionResolverInterface`.

This has practical implications, because that argument is injected by the service container. I was replacing the binding for `ConnectionResolverInterface` with my own implementation and it was causing issues at runtime. If what is really necessary is a `DatabaseManager`, that should be the type of that argument, so that the service container can provide the proper instance.

As someone trying to implement a new database paradigm (NoSQL), I would prefer to fix the interfaces instead. But as you said, it's very possible that adding new connections in that way is not intended at the moment. In that case maybe the interfaces should be removed altogether to prevent confusions.

**Comment by RemiCollin** [comment]
I totally agree it has to be improved if we want to add custom DB implementations; And as a mongoDB user, I'm interested in your project, as the current popular solution really feels abandoned by its maintainer.

**Comment by NoelDeMartin** [comment]
Regardless of supporting MongoDB or not, there is a design problem here. Maybe this doesn't seem too dramatic because we are talking about migrations. But I've also found the same problem in the eloquent model.

Eloquent models have the method `setConnectionResolver` accepting a `ConnectionResolverInterface`. And that trickles down to the method `newBaseQueryBuilder` (which is practically used every time eloquent interacts with the DB). In that method, it is calling the connection methods `getQueryGrammar` and `getPostProcessor` that do not exist on the `ConnectionInterface` interface.

So, it is not a stretch to say that this is happening in most Laravel requests interacting with the DB at some point.
